gcp_variant_transforms/beam_io/vcfio.py

```diff
@@ -207,6 +209,156 @@ def __repr__(self):
+
+    def _get_variant_info(self, variant):
+      info = []
+      # Set END in info if it doesn't match start+len(reference_bases). This is
+      # usually the case for non-variant regions.
+      if (variant.start is not None
+          and variant.reference_bases
+          and variant.end
+          and variant.start + len(variant.reference_bases) != variant.end):
+        info.append('END=%d' % variant.end)
+      # Set all other fields of info.
+      for k, v in variant.info.iteritems():
+        if v.data is True:
+          info.append(k)
+        else:
+          info.append('%s=%s' % (k, self._format_value(v.data)))
+      return info
+
+    def _get_variant_format_keys(self, variant):
+      if not variant.calls:
+        return []
+
+      format_keys = [GENOTYPE_FORMAT_KEY]
+      for call in variant.calls:
+        # If any calls have a set phaseset that is not `DEFAULT_PHASESET_VALUE`,
+        # the key will be added to the format field.
+        if self._is_alternate_phaseset(call):
+          format_keys.append(PHASESET_FORMAT_KEY)
+        format_keys.extend([k for k in call.info])
+
+      # Sort all keys and remove duplicates after GT
+      format_keys[1:] = sorted(list(set(format_keys[1:])))
+
+      return format_keys
+
+    def _get_variant_calls(self, variant, format_keys):
+      calls = []
+      for call in variant.calls:
+        call_info = []
+        # GT will always be the first format key.
+        genotype = []
+        for allele in call.genotype:
+          if allele == MISSING_GENOTYPE_VALUE:
+            genotype.append(MISSING_FIELD_VALUE)
+          else:
+            genotype.append(self._format_value(allele))
+        phase_char = '|' if call.phaseset else '/'
+        call_info.append(phase_char.join(genotype))
+
+        for key in format_keys[1:]:
+          if key == PHASESET_FORMAT_KEY:
+            if self._is_alternate_phaseset(call):
+              call_info.append(call.phaseset)
+            else:
+              call_info.append(MISSING_FIELD_VALUE)
+          else:
+            call_info.append(self._format_value(
+                call.info[key] if key in call.info else None))
+
+        calls.append(':'.join(call_info))
+
+      return calls
+
+    def encode(self, variant):
+      """Converts a :class:`Variant` object back to a VCF line."""
+      info = self._get_variant_info(variant)
+      format_keys = self._get_variant_format_keys(variant)
+      calls = self._get_variant_calls(variant, format_keys)
+
+      columns = [
+          variant.reference_name,
+          variant.start if variant.start is None else variant.start + 1,
+          ';'.join(variant.names),
+          variant.reference_bases,
+          ','.join(variant.alternate_bases),
+          variant.quality,
+          ';'.join(variant.filters),
+          ';'.join(info),
+          ':'.join(format_keys),
+      ]
+      if calls:
+        columns.append('\t'.join(calls))
+
+      for i, column in enumerate(columns):
```

Comment:
would "columns = [self._format_value(c) for c in columns]" also work?

Reply:
Yes, definitely better!